firmware/src/pwm.c

```diff
@@ -30,3 +32,51 @@ void pwm_init(void)
     set_bit(PWM_DDR, PWM);                          // PWM as output
 
 }
+void pwm_reset(void)
+{
+    set_pwm_off();
+#ifdef MACHINE_ON
+    control.D = 0;
+#endif
+    usart_send_string("PWM turned off!\n");
+}
+void pwm_compute(void){
+#ifdef MACHINE_ON
+    
+    if(!control.sweep_done){
+        sweep_duty();
+    }else{
+        perturb_and_observe();
+    }

```

Comment:
its a bit confusing have this here, I recommend to have a control separated file that does the control logic and call sweep or duty.

and in pwm file **only** add code related with pwm